astropy/modeling/astro_sherpa.py

```diff
@@ -0,0 +1,433 @@
+    x = np.array(x)
+    y = np.array(y)
+
+    if x.ndim == 2 or (x.dtype == np.object or y.dtype == np.object):
+        data = []
+        if z is None:
+            z = len(x) * [None]
+
+        if xerr is None:
+            xerr = len(x) * [None]
+
+        if yerr is None:
+            yerr = len(y) * [None]
+
+        if zerr is None:
+            zerr = len(z) * [None]
+
+        for nn, (xx, yy, zz, xxe, yye, zze) in enumerate(zip(x, y, z, xerr, yerr, zerr)):
+            data.append(_make_dataset(n_dim, x=xx, y=yy, z=zz, xerr=xxe, yerr=yye, zerr=zze, n=nn))
+        return DataSimulFit("wrapped_data", data)
+    else:
+        return _make_dataset(n_dim, x=x, y=y, z=z, xerr=xerr, yerr=yerr, zerr=zerr)
+
+
+def _make_simfit_model(models, tie_list, model_dict):
+    '''
+    makes a SimulFitModel for the models suplied it also links and parameters within tie_list
+    model_dict updated with astropy models and their converted conterparts this is used for
+    repopulating the model later
+    '''
+    for mod in models:
+        model_dict[mod] = _astropy_to_sherpa_model(mod)
+
+    if tie_list is not None:
+        for par1, par2 in tie_list:
+            getattr(model_dict[par1._model], par1.name).link = getattr(model_dict[par2._model], par2.name)
+    return SimulFitModel("wrapped_fit_model", model_dict.values())
+
+
+class SherpaFitter(Fitter):
+
+    """
+    Sherpa Fitter for astropy models. Yay :)
+
+    Parameters
+        ----------
+        optimizer : (callable)
+                    A sherpa optimizer method.
+        statistic : (callable)
+                    A sherpa statistic function.
+        estmethod : (callable)
+                    A shepra error estimation method.
+    """
+
+    def __init__(self, optimizer=LevMar, statistic=LeastSq, estmethod=Covariance):
+        '''
+        check the types of optimizer, statistic and estmethod are sherpa based classes
+         so we can use them
+        '''
+        if inspect.isclass(optimizer) and not issubclass(optimizer, OptMethod):
+            raise TypeError("optimizer must be sherpa OptMethod class or isinstance")
+        elif inspect.isfunction(optimizer) and not issubclass(optimizer.__class__, OptMethod):
+            raise TypeError("optimizer must be sherpa OptMethod class or isinstance")
+
+        if inspect.isclass(statistic) and not issubclass(statistic, Stat):
+            raise TypeError("statistic must be sherpa Stat class or isinstance")
+        elif inspect.isfunction(statistic) and not issubclass(statistic.__class__, Stat):
+            raise TypeError("statistic must be sherpa Stat class or isinstance")
+
+        if inspect.isclass(estmethod) and not issubclass(estmethod, EstMethod):
+            raise TypeError("estmethod must be sherpa EstMethod class or isinstance")
+        elif inspect.isfunction(estmethod) and not issubclass(estmethod.__class__, EstMethod):
+            raise TypeError("estmethod must be sherpa EstMethod class or isinstance")
```

Comment:
I don't think you need all of this strictness. I think we agreed we should provide an API where you can use strings for selecting optimizers, statistics, etc. We could still provide a way for users to provide their own implementations: I am not sure how worthwhile that is right now, but even in that case I would not worry too much about the instance's identity and I would trust the user to provide an implementation of the correct contract, which should be documented elsewhere.

So, if we accept strings, you could just have three factory methods that build instances of the relevant Sherpa classes, raise an error if there is something wrong with the string, and move on.

However, users should also be able to pass options to the optimization method. Maybe we should take a step back, agree on the definition of the API and then refactor this code accordingly.